statsmodels/tsa/vecm/vecm.py

```diff
@@ -0,0 +1,254 @@
+        K = y_all.shape[0]
+        
+        y_1_T = y_all[:, p:]
+        T = y_1_T.shape[1]
+        
+        Delta_y_all = np.diff(y_all)
+        Delta_y_1_T = Delta_y_all[:, p-1:]
+        
+        y_minus1 = y_all[:, p-1:-1]
+        if 'lt' in deterministic_terms:
+            y_minus1 = np.vstack((y_minus1, np.arange(T)))
+        
+        Delta_x = np.zeros((diff_lags*K, T))
+        for j in range(Delta_x.shape[1]):
+            Delta_x[:, j] = Delta_y_all[:, j+p-2:None if j-1<0 else j-1:-1].T.reshape(K*(p-1))
+        if 'c' in deterministic_terms:
+            Delta_x = np.vstack((Delta_x, np.ones(T)))
+        if 's' in deterministic_terms: # TODO: How many seasons??
+            seasons = np.zeros((3, Delta_x.shape[1])) # 3 = 4-1 (4=number_of_seasons)
+            for i in range(seasons.shape[0]):
+                seasons[i, i::4] = i+1
+            Delta_x = np.vstack((Delta_x, seasons))
+        
+        return y_all, y_1_T, Delta_y_all, Delta_y_1_T, y_minus1, Delta_x
+
+    def _ls_Pi_Gamma(self, Delta_y_1_T, y_minus1, Delta_x, deterministic_terms):
+        mat1 = np.hstack((Delta_y_1_T.dot(y_minus1.T), Delta_y_1_T.dot(Delta_x.T)))
+
+        B = y_minus1.dot(Delta_x.T)
+        mat2 = np.linalg.inv(np.vstack( (np.hstack( (y_minus1.dot(y_minus1.T),   B                    ) ),
+                                         np.hstack( (B.T,                        Delta_x.dot(Delta_x.T) )) ) ))
+
+        estPiGamma = mat1.dot(mat2)
+
+        Pi_hat_cols = K if 'lt' not in deterministic_terms else K+1
+        Pi_hat = estPiGamma[:, :Pi_hat_cols]
+
+        Gamma_hat = estPiGamma[:, Pi_hat_cols:]
+        #print("Dy1T.shape: "+str(Delta_y_1_T.shape) + ",  Pi_hat.shape: "+str(Pi_hat.shape) + ",  Gamma_hat.shape: "+str(Gamma_hat.shape))
+        #print("yminus1:" + str(y_minus1.shape))
+        _A = Delta_y_1_T - Pi_hat.dot(y_minus1) - Gamma_hat.dot(Delta_x)
+        p = diff_lags+1
+        Sigma_u_hat = 1/(T-K*p) * np.dot(_A,_A.T)
+
+        return (Pi_hat, Gamma_hat, Sigma_u_hat)
+
+    def _estimate_vecm_ls(self, diff_lags, deterministic_terms):    
+        # deterministic_terms \in \{'c', 'lt', 's'\} c=constant, lt=linear trend, s=seasonal terms
+
+        y_all, y_1_T, Delta_y_all, Delta_y_1_T, y_minus1, Delta_x = self.estimation_matrices(diff_lags, deterministic_terms)
+        K = y_all.shape[0]
+        T = y_1_T.shape[1]
+
+        Pi_hat, Gamma_hat, Sigma_u_hat = _ls_Pi_Gamma(Delta_y_1_T, y_minus1, Delta_x, deterministic_terms)
+        
+        print("Pi:")
+        print(Pi_hat)
+        print("Gamma:")
+        print(Gamma_hat)
+        print("Sigma_u")
+        print(Sigma_u_hat)
+        
+        return {"Pi_hat": Pi_hat, "Gamma_hat": Gamma_hat, "Sigma_u_hat": Sigma_u_hat}        
+
+    def _M_and_R_matrices(self, T, Delta_x, Delta_y_1_T, y_minus1):
+        M = np.identity(T) - Delta_x.T.dot(np.linalg.inv(Delta_x.dot(Delta_x.T))).dot(Delta_x)
```

Comment:
At the end of refactoring, there should be almost no `inv` left. I usually prefer the full path for the few occasions where we need them.

note M will have to go away completely because it's nobs x nobs, and be replace with a multivariate residual function. 
I think after unit test coverage is sufficiently high and tests pass, we can streamline the linear algebra from the textbook version to computationally more efficient versions.